src/utils.rs

```diff
@@ -38,52 +38,41 @@ where
     }
 }
 
-pub fn expand_trns_line(buf: &mut [u8], trns: &[u8], channels: usize) {
-    // Return early if empty. This enables to subtract `channels` later without overflow.
-    if buf.len() < (channels + 1) {
-        return;
+pub fn expand_trns_line(input: &[u8], output: &mut [u8], trns: &[u8], channels: usize) {
+    for (input, output) in input
+        .chunks_exact(channels)
+        .zip(output.chunks_exact_mut(channels + 1))
+    {
+        output[..channels].copy_from_slice(input);
+        output[channels] = if input == trns { 0 } else { 0xFF };
     }
+}
 
-    let i = (0..=buf.len() / (channels + 1) * channels - channels)
-        .rev()
-        .step_by(channels);
-    let j = (0..=buf.len() - (channels + 1)).rev().step_by(channels + 1);
-    for (i, j) in i.zip(j) {
-        let i_pixel = i;
-        let j_chunk = j;
-        if &buf[i_pixel..i_pixel + channels] == trns {
-            buf[j_chunk + channels] = 0
+pub fn expand_trns_line16(input: &[u8], output: &mut [u8], trns: &[u8], channels: usize) {
+    for (input, output) in input
+        .chunks_exact(channels * 2)
+        .zip(output.chunks_exact_mut(channels * 2 + 2))
+    {
+        output[..channels * 2].copy_from_slice(input);
+        if input == trns {
+            output[channels * 2] = 0;
+            output[channels * 2 + 1] = 0
```

Comment:
For expanding trns, you can take advantage of knowing that you only expand color components for Luma and RGB to their equivalents with alpha.

So code can look like (for 16 bit trns)

```Rust
use std::convert::TryInto;

pub fn expand_trns_line16(input: &[u8], output: &mut [u8], trns: &[u8], channels: usize) {
    
    match channels {
        3 => {
            let trns_bytes: &[u8; 6] = trns.try_into().unwrap();

            for (input, output) in input.chunks_exact(6).zip(output.chunks_exact_mut(8)) {
                output[..6].copy_from_slice(input);

                if input == trns_bytes {
                    output[6] = 0;
                    output[7] = 0;
                } else {
                    output[6] = 0xFF;
                    output[7] = 0xFF;
                };
            }
        }

        1 => {
            let trns_bytes: &[u8; 2] = trns.try_into().unwrap();

            for (input, output) in input.chunks_exact(2).zip(output.chunks_exact_mut(4)) {
                output[..2].copy_from_slice(input);

                if input == trns_bytes {
                    output[2] = 0;
                    output[3] = 0;
                } else {
                    output[2] = 0xFF;
                    output[3] = 0xFF;
                };
            }
        }

        _ => unreachable!(),
    }
}

````

Godbolt: https://godbolt.org/z/oh6PE9717